ClassCastException when deserializing data with an extra @ForyField whose type is already registered in consumer

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

- Fory version: 0.15.0
- JDK: 21
- Language: Java

### Component(s)

Java

### Minimal reproduce step

When using Fory with withCompatible(true) and @ForyField annotations, adding a new field to a class breaks backward compatibility. Specifically, when the producer (new code with the extra field) serializes data, and the consumer (old code without the extra field) tries to deserialize it, a ClassCastException is thrown.

The issue occurs when the extra field's type is a class that is already registered in the consumer's Fory instance (e.g., the consumer already has another field of the same type).

```java
import lombok.Data;
import org.apache.fory.Fory;
import org.apache.fory.config.Language;
import org.apache.fory.annotation.ForyField;

// ---- Shared class, registered in both producer and consumer ----

@Data
public class Address {
    @ForyField(id = 1, nullable = true)
    private String city;

    @ForyField(id = 2, nullable = true)
    private String zip;
}

// ---- Consumer version: 3 fields, already has an Address field ----

@Data
public class UserV1 {
    @ForyField(id = 1)
    private Long id;

    @ForyField(id = 2, nullable = true)
    private String name;

    @ForyField(id = 3, nullable = true)
    private Address homeAddress;  // Address is already registered
}

// ---- Producer version: 4 fields, added another Address field ----

@Data
public class UserV2 {
    @ForyField(id = 1)
    private Long id;

    @ForyField(id = 2, nullable = true)
    private String name;

    @ForyField(id = 3, nullable = true)
    private Address homeAddress;

    @ForyField(id = 4, nullable = true)
    private Address workAddress;  // new field, same type as homeAddress
}
``` 

```java
// ---- Producer Fory: registers UserV2 + Address ----
Fory producerFory = Fory.builder()
        .withLanguage(Language.JAVA)
        .withCompatible(true)
        .requireClassRegistration(true)
        .build();
producerFory.register(UserV2.class);
producerFory.register(Address.class);

// ---- Consumer Fory: registers UserV1 + Address (old code) ----
Fory consumerFory = Fory.builder()
        .withLanguage(Language.JAVA)
        .withCompatible(true)
        .requireClassRegistration(true)
        .build();
consumerFory.register(UserV1.class);
consumerFory.register(Address.class);

// ---- Serialize with producer (new code) ----
UserV2 user = new UserV2();
user.setId(1L);
user.setName("Alice");

Address home = new Address();
home.setCity("NYC");
home.setZip("10001");
user.setHomeAddress(home);

Address work = new Address();   // non-null value for the new field
work.setCity("SF");
work.setZip("94000");
user.setWorkAddress(work);

byte[] bytes = producerFory.serialize(user);

// ---- Deserialize with consumer (old code) — throws ClassCastException ----
UserV1 result = (UserV1) consumerFory.deserialize(bytes);
``` 




### What did you expect to see?

With withCompatible(true), the consumer (old code with 3 fields) should be able to deserialize data serialized by the producer (new code with 4 fields). The extra workAddress field should be skipped (read and discarded), and the id, name, and homeAddress fields should be correctly deserialized.

### What did you see instead?

org.apache.fory.exception.DeserializationException: Failed to deserialize input
    at org.apache.fory.Fory.deserialize(Fory.java:893)
    at org.apache.fory.Fory.deserialize(Fory.java:807)
    ...

Caused by: java.lang.ClassCastException: class Address cannot be cast to class org.apache.fory.serializer.UnknownClass$UnknownStruct
    at UserV2ForyCodecMetaShared4_0.readFields2$(UserV2ForyCodecMetaShared4_0.java:149)
    at UserV2ForyCodecMetaShared4_0.read(UserV2ForyCodecMetaShared4_0.java:162)
    at org.apache.fory.Fory.readDataInternal(Fory.java:1065)
    at org.apache.fory.Fory.deserialize(Fory.java:889)
    ...

### Anything Else?

- The issue does not occur when the extra field's value is null.
- The issue is not specific to the Address type — any registered class used as the new field type triggers the same error.
- This is a critical issue for rolling deployments: when a new field is added (even reusing an existing registered type), all services that deserialize the data must be updated simultaneously, which defeats the purpose of withCompatible(true).

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fory-core/src/main/java/org/apache/fory/builder/CompatibleCodecBuilder.java
+++ b/java/fory-core/src/main/java/org/apache/fory/builder/CompatibleCodecBuilder.java
@@ -343,6 +343,9 @@ public class CompatibleCodecBuilder extends ObjectCodecBuilder {
   protected Expression deserializeField(
       Expression buffer, Descriptor descriptor, Function<Expression, Expression> callback) {
     FieldConverter<?> converter = descriptor.getFieldConverter();
+    if (shouldSkipField(descriptor)) {
+      return skipField(descriptor);
+    }
     if (converter == null) {
       return super.deserializeField(buffer, descriptor, callback);
     }
@@ -351,6 +354,36 @@ public class CompatibleCodecBuilder extends ObjectCodecBuilder {
         targetValue != null,
         "Unsupported compatible scalar converter target " + FieldConverters.toType(converter));
     return new Expression.ListExpression(targetValue, callback.apply(targetValue));
+  }
+
+  @Override
+  protected Expression deserializeGroupForRecord(
+      List<Descriptor> group, Expression bean, Expression buffer) {
+    Expression.ListExpression expressions = new Expression.ListExpression();
+    for (Descriptor descriptor : group) {
+      if (shouldSkipField(descriptor)) {
+        expressions.add(skipField(descriptor));
+        continue;
+      }
+      TypeRef<?> castTypeRef = readValueTypeRef(descriptor);
+      Expression value = deserializeField(buffer, descriptor, expression -> expression);
+      expressions.add(setFieldValue(bean, descriptor, tryInlineCast(value, castTypeRef)));
+    }
+    return expressions;
+  }
+
+  private boolean shouldSkipField(Descriptor descriptor) {
+    return descriptor.getField() == null
+        && descriptor.getFieldConverter() == null
+        && !descriptor.getRawType().isPrimitive();
+  }
+
+  private Expression skipField(Descriptor descriptor) {
+    return new Expression.Invoke(
+        new Expression.Reference("this", TypeRef.of(GeneratedCompatibleSerializer.class), false),
+        "skipField",
+        readContextRef(),
+        remoteFieldInfo(descriptor));
   }
 
   @Override

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/AbstractObjectSerializer.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/AbstractObjectSerializer.java
@@ -207,6 +207,18 @@ public abstract class AbstractObjectSerializer<T> extends Serializer<T> {
     return null;
   }
 
+  protected final void skipField(ReadContext readContext, SerializationFieldInfo remoteFieldInfo) {
+    // A remote-only struct can have a synthetic UnknownStruct descriptor while registered type
+    // dispatch still materializes its concrete local class. Generated compatible readers must use
+    // this untyped path so consuming the field never inserts a cast to the synthetic type.
+    FieldSkipper.skipField(
+        readContext,
+        typeResolver,
+        readContext.getRefReader(),
+        remoteFieldInfo,
+        readContext.getBuffer());
+  }
+
   /**
    * Write field value to buffer by reading from the object via fieldAccessor. Handles primitive
    * types, unsigned/compressed numbers, and common types like String with optimized fast paths.

--- a/java/fory-core/src/test/java/org/apache/fory/serializer/MetaShareCompatibleTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/serializer/MetaShareCompatibleTest.java
@@ -32,10 +32,13 @@ import java.util.stream.Collectors;
 import org.apache.fory.Fory;
 import org.apache.fory.ForyTestBase;
 import org.apache.fory.TestUtils;
+import org.apache.fory.annotation.ForyField;
+import org.apache.fory.annotation.Nullable;
 import org.apache.fory.builder.CompatibleCodecBuilder;
 import org.apache.fory.codegen.CompileUnit;
 import org.apache.fory.codegen.JaninoUtils;
 import org.apache.fory.config.ForyBuilder;
+import org.apache.fory.config.Language;
 import org.apache.fory.context.MetaReadContext;
 import org.apache.fory.context.MetaWriteContext;
 import org.apache.fory.meta.NativeTypeDefEncoderTest;
@@ -90,6 +93,46 @@ public class MetaShareCompatibleTest extends ForyTestBase {
       this.id = id;
       this.price = price;
     }
+  }
+
+  public static final class Address {
+    @Nullable
+    @ForyField(id = 1)
+    public String city;
+
+    @Nullable
+    @ForyField(id = 2)
+    public String zip;
+  }
+
+  public static final class UserV1 {
+    @ForyField(id = 1)
+    public Long id;
+
+    @Nullable
+    @ForyField(id = 2)
+    public String name;
+
+    @Nullable
+    @ForyField(id = 3)
+    public Address homeAddress;
+  }
+
+  public static final class UserV2 {
+    @ForyField(id = 1)
+    public Long id;
+
+    @Nullable
+    @ForyField(id = 2)
+    public String name;
+
+    @Nullable
+    @ForyField(id = 3)
+    public Address homeAddress;
+
+    @Nullable
+    @ForyField(id = 4)
+    public Address workAddress;
   }
 
   @DataProvider
@@ -856,6 +899,52 @@ public class MetaShareCompatibleTest extends ForyTestBase {
       Object newObj = fory.deserialize(serialized);
       System.out.println(newObj);
     }
+  }
+
+  @Test(dataProvider = "enableCodegen")
+  public void testRegisteredMissingStruct(boolean enableCodegen) {
+    int userTypeId = 100;
+    int addressTypeId = 101;
+    Fory writer =
+        Fory.builder()
+            .withLanguage(Language.JAVA)
+            .withCompatible(true)
+            .withCodegen(enableCodegen)
+            .withAsyncCompilation(false)
+            .requireClassRegistration(true)
+            .build();
+    writer.register(UserV2.class, userTypeId);
+    writer.register(Address.class, addressTypeId);
+
+    UserV2 user = new UserV2();
+    user.id = 1L;
+    user.name = "Alice";
+    Address homeAddress = new Address();
+    homeAddress.city = "NYC";
+    homeAddress.zip = "10001";
+    user.homeAddress = homeAddress;
+    Address workAddress = new Address();
+    workAddress.city = "SF";
+    workAddress.zip = "94000";
+    user.workAddress = workAddress;
+    byte[] bytes = writer.serialize(user);
+
+    Fory reader =
+        Fory.builder()
+            .withLanguage(Language.JAVA)
+            .withCompatible(true)
+            .withCodegen(enableCodegen)
+            .withAsyncCompilation(false)
+            .requireClassRegistration(true)
+            .build();
+    reader.register(UserV1.class, userTypeId);
+    reader.register(Address.class, addressTypeId);
+
+    UserV1 result = (UserV1) reader.deserialize(bytes);
+    Assert.assertEquals(result.id, Long.valueOf(1));
+    Assert.assertEquals(result.name, "Alice");
+    Assert.assertEquals(result.homeAddress.city, "NYC");
+    Assert.assertEquals(result.homeAddress.zip, "10001");
   }
 
   @Test

--- a/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
+++ b/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
@@ -30,7 +30,10 @@ import java.util.Arrays;
 import java.util.List;
 import org.apache.fory.Fory;
 import org.apache.fory.ThreadSafeFory;
+import org.apache.fory.annotation.ForyField;
+import org.apache.fory.annotation.Nullable;
 import org.apache.fory.config.ForyBuilder;
+import org.apache.fory.config.Language;
 import org.apache.fory.context.MetaReadContext;
 import org.apache.fory.context.MetaWriteContext;
 import org.apache.fory.test.bean.Struct;
@@ -48,6 +51,20 @@ public class RecordSerializersTest {
 
   public record BoxedPrimitiveRecord(String lensId, Long from, Long to, Integer type)
       implements Serializable {}
+
+  public record AddressRecord(
+      @Nullable @ForyField(id = 1) String city, @Nullable @ForyField(id = 2) String zip) {}
+
+  public record UserRecordV1(
+      @ForyField(id = 1) Long id,
+      @Nullable @ForyField(id = 2) String name,
+      @Nullable @ForyField(id = 3) AddressRecord homeAddress) {}
+
+  public record UserRecordV2(
+      @ForyField(id = 1) Long id,
+      @Nullable @ForyField(id = 2) String name,
+      @Nullable @ForyField(id = 3) AddressRecord homeAddress,
+      @Nullable @ForyField(id = 4) AddressRecord workAddress) {}
 
   @Test
   public void testIsRecord() {
@@ -91,6 +108,40 @@ public class RecordSerializersTest {
             .build();
     Foo foo = new Foo(10, "abc", new ArrayList<>(Arrays.asList("a", "b")), 'x');
     Assert.assertEquals(fory.deserialize(fory.serialize(foo)), foo);
+  }
+
+  @Test(dataProvider = "codegen")
+  public void testRegisteredMissingStruct(boolean codegen) {
+    int userTypeId = 100;
+    int addressTypeId = 101;
+    Fory writer =
+        Fory.builder()
+            .withLanguage(Language.JAVA)
+            .withCompatible(true)
+            .withCodegen(codegen)
+            .withAsyncCompilation(false)
+            .requireClassRegistration(true)
+            .build();
+    writer.register(UserRecordV2.class, userTypeId);
+    writer.register(AddressRecord.class, addressTypeId);
+
+    AddressRecord homeAddress = new AddressRecord("NYC", "10001");
+    AddressRecord workAddress = new AddressRecord("SF", "94000");
+    byte[] bytes = writer.serialize(new UserRecordV2(1L, "Alice", homeAddress, workAddress));
+
+    Fory reader =
+        Fory.builder()
+            .withLanguage(Language.JAVA)
+            .withCompatible(true)
+            .withCodegen(codegen)
+            .withAsyncCompilation(false)
+            .requireClassRegistration(true)
+            .build();
+    reader.register(UserRecordV1.class, userTypeId);
+    reader.register(AddressRecord.class, addressTypeId);
+
+    UserRecordV1 result = (UserRecordV1) reader.deserialize(bytes);
+    Assert.assertEquals(result, new UserRecordV1(1L, "Alice", homeAddress));
   }
 
   @Test

